# Introduction to Quantum Computing
<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/google-quantum.png"
     alt="GOOGLE-QUANTUM"
     width="25%">
<p>(Cho, 2023)

### Prof. David Singletary
### Florida State College at Jacksonville

# Day 2: Quantum Gates and Circuits
- Hands-on: Measuring Qubits and Interpreting Results (Knapsack Problem)
- More on Single-Qubit States and Gates (Ch. 3)
- Hands-on: A Bloch Sphere Visualization Tool
- Hands-On: Preparing a Single Qubit State
- Common Single-Qubit Gates
-	Hands-on: Building Simple Quantum Circuits
-	Hands-on: Simulating Circuits and Analyzing Outcomes


# Hands-On: The Knapsack Problem
(Gonciulea & Stefanski, 2025, p. 20)

The knapsack problem is a classic example of a binary optimization problem.
- Given a set of items, each with a weight and a value, and a knapsack with a maximum weight capacity, select the combination of items that maximizes total value without exceeding the weight limit.
- Each item represents a binary decision: include it (1) or exclude it (0).

- In this example we solve a small knapsack problem with a maximum weight capacity of 4 units and three items weighing 3, 2, and 1 units.
- Their values are \$2000, \$3000, and \$1000 (scaled down to 2, 3, and 1 for simplicity).
- The goal is to evaluate all possible item combinations and determine which combination maximizes value without exceeding the weight limit.

**Table 2.1 Items, Weights, and Values for the Knapsack Problem**

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-table2.1.png"
      alt="Knapsack Problem Table 2.1"
      width="60%">

Can you work out the optimal solution in your head?

## Scaling the Problem

With n binary variables, there are 2ⁿ possible combinations. While small examples are manageable, the number of combinations grows exponentially, becoming enormous in real-world problems.
- For example, 41 items lead to over one trillion possible combinations, 2^41=2,199,023,255,552.



## Evaluating the Possible Selections

**Table 2.2 Possible Item Selection with Corresponding Value and Weight**

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-table2.2.png"
      alt="Knapsack Problem Table 2.2"
      width="75%">

## Representing Outcomes with Binary Strings
- There are 2n binary strings with length n, where n > 0 is the number of digits.
- We often talk about measurement outcomes using decimal (base 10 integers)
- The following tables show the decimal (base 10) values
of binary strings of lengths one, two, and three.

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/binary-decimal.png"
      alt="Binary/Decimal Strings"
      width="75%">

- Each qubit in a quantum system corresponds to one binary digit in the measurement outcome.
- A quantum system with n qubits has 2n possible outcomes, so for this problem there are 8 (2^3) possible outcomes.

## Evaluating All Combinations as Binary Strings  (sequentially ordered):

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-evalcombinations.png"
      alt="Knapsack Problem Combinations"
      width="50%">

Optimal solution: 011 (Items 1 and 2)
Total weight: 3
Total value: 4

## Quantum state and probabilities

Each run of a quantum computation produces a single outcome.
- By executing a computation multiple times, measured outcomes are collected that reveal the underlying probability distribution.

  **Rose Chart Representation of 3-Qubit System**

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-figure2.5.png"
      alt="3-bit Rose Chart"
      width="50%">

- The rose chart representation is an intuitive way to visualize the setup of a quantum computation, but in many cases, it is useful to “unwrap” the wheel into a **state table**, where we can add columns to include more information.

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-figure2.6.png"
      alt="3 Qubit System Tabular Representation"
      width="25%">

As with the sectors of the rose chart, the probability is represented by the length of the bar relative to the total length of the column.

## What Does a Single Run (Shot) Look Like?

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-figure2.2.png"
      alt="Knapsack Problem Figure 2.2"
      width="125%">

A single run of a quantum computation involves preparing the qubits in an initial state, applying a sequence of quantum operations, and then measuring the system to produce one outcome.
- “counts” in this figure refers to the number of times each measurement outcome appears after running the quantum circuit many times.
- By repeating this process many times, we gather a distribution of results.
- **"The goal for many quantum computations is to increase the
probability of measuring desired outcomes."**

- Similar to how classical programs use registers or to store values, **quantum registers** are groups of qubits used to represent different variables in a computation.
- In the knapsack problem, separate registers are used to track item selection, total value, and total weight.

- Three registers encoding selection, value, and weight
  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/knapsack-figure2.4.png"
      alt="Three registers for Knapsack Problem"
      width="75%">

  - The number of qubits in each register depends on how much information must be represented for the problem.
  - The selection register uses one qubit per item to represent whether each item is included (1) or not (0).
  - Additional registers store the total value and weight, which are calculated as functions of the binary selection variables.
  - This structured encoding prepares the problem for implementation within a quantum circuit.

The knapsack problem can be solved using a solve_knapsack function that takes item values, weights, and a maximum weight limit as inputs.
- The algorithm uses two key quantum patterns: an oracle to recognize valid selections and amplitude amplification to increase the probability of desirable outcomes.
- It begins by identifying selections that meet the weight constraint and a minimum value threshold, then iteratively raises the minimum value while amplifying valid solutions.
- Through repeated measurement and refinement, the algorithm ultimately returns the optimal selection that maximizes value without exceeding the weight limit.


In [1]:
# ---- One-time Colab setup ----

import os
import sys
import subprocess
import importlib

REPO_URL = "https://github.com/learnqc/code.git"
REPO_DIR = "/content/code"
SRC_DIR = f"{REPO_DIR}/src"

# Remove the unrelated PyPI hume package if it exists
subprocess.run(
    ["pip", "uninstall", "-y", "hume"],
    stdout=subprocess.DEVNULL,
    stderr=subprocess.DEVNULL
)

# Install required pip package(s)
subprocess.run(
    ["pip", "install", "-q", "sty"],
    check=True
)

# Clone repo if needed
if not os.path.exists(REPO_DIR):
    subprocess.run(
        ["git", "clone", REPO_URL, REPO_DIR],
        check=True
    )

# Make src importable
if SRC_DIR not in sys.path:
    sys.path.insert(0, SRC_DIR)

# Clear any stale imports
importlib.invalidate_caches()

# Verify imports
from hume import QuantumRegister, QuantumCircuit
from hume.utils.common import padded_bin, complex_to_rgb
from hume.algos.grover_optimizer import grover_optimizer, oracle_match_0_multi
from hume.algos.function_encoding import encode_term

print("Setup complete. Hume imports are working.")

Setup complete. Hume imports are working.


In [2]:
from math import floor, log10, atan2, log2, pi

from sty import RgbBg, Style, bg, fg

from hume.utils.common import padded_bin
from hume.utils.common import complex_to_rgb

# Select the simulation backend: 'hume' uses the Hume educational quantum
# simulator, while the commented line can be enabled to use the Qiskit backend.
# The Hume simulator is designed specifically for learning and visualization,
# rather than performance or hardware compatibility. It exposes the internal
# quantum state directly, allowing users to see how amplitudes change during
# a computation.
SIMULATOR = 'hume'
# SIMULATOR = 'qiskit'

# define a function that formats and displays a quantum knapsack state as a
# table of decoded register values, amplitudes, phases, and amplitude bars.
def knapsack_table_new(
    state,
    key_size,
    value_size,
    weight_size,
    decimals=4,
    symbol='\u2588',
    colors=None
):
    n = int(log2(len(state)))
    round_state = [
        round(state[k].real, decimals)
        + 1j * round(state[k].imag, decimals)
        for k in range(len(state))
    ]

    headers = [
        'Selection',
        'Value',
        'Binary',
        'Weight',
        'Binary',
        'Magnitude',
        'Direction',
        'Amplitude Bar'
    ]
    offsets = [
        # max(len(headers[0]), floor(log10(len(state)))),  # outcome key
        max(len(headers[0]), key_size),                   # binary key
        max(len(headers[1]), floor(log10(len(state)))),   # outcome value
        max(len(headers[2]), value_size),                 # binary value
        max(len(headers[3]), floor(log10(len(state)))),   # outcome weight
        max(len(headers[4]), n - key_size - value_size),  # binary weight
        # max(len(headers[4]), 2 * (decimals + 2) + 6),   # amplitude
        max(len(headers[5]), decimals + 2),               # magnitude
        max(len(headers[6]), decimals),                   # direction
        max(len(headers[7]), 24),                         # amplitude bar
        # max(len(headers[8]), decimals + 2),             # probability
    ]

    for i in range(len(offsets)):
        headers[i] = headers[i] + ' ' * (offsets[i] - len(headers[i]))

    header_str = '  '.join(headers)

    output = '\n'
    # output += '\033[1m' + '      Key             Value' + '\033[0m'
    # output += '\n'
    output += header_str
    output += '\n'
    output += len(header_str) * '-'
    output += '\n'

    ind_to_color = []  # array containing the color for each row index

    # populate ind_to_color with None for each row
    for i in range(len(round_state)):
        ind_to_color.append(None)

    # add the color for each index that's supposed to be colored
    if colors is not None:
        for key in colors:
            for index in colors.get(key):
                # make sure each index is only assigned one color
                assert(not ind_to_color[index])
                ind_to_color[index] = key

    rows = []
    for k in range(len(round_state)):
        if abs(round_state[k]) > 0.0001:
            direction = round(
                atan2(round_state[k].imag, round_state[k].real) * 180 / pi,
                2
            )

            # if creating a row that has a background color, get the color
            if ind_to_color[k]:
                bg.my_color = Style(
                    RgbBg(
                        ind_to_color[k][0],
                        ind_to_color[k][1],
                        ind_to_color[k][2]
                    )
                )

            rows += [[
                padded_bin(n, k)[-key_size:].ljust(offsets[0], ' '),

                str(
                    int(padded_bin(n, k)[weight_size:-key_size], 2)
                ).ljust(offsets[1], ' '),

                padded_bin(n, k)[weight_size:-key_size].ljust(
                    offsets[2],
                    ' '
                ),

                str(
                    int(padded_bin(n, k)[0:weight_size], 2)
                ).ljust(offsets[3], ' '),

                padded_bin(n, k)[0:weight_size].ljust(offsets[4], ' '),

                str(round(abs(state[k]), decimals)).ljust(
                    decimals + 2,
                    ' '
                ).ljust(offsets[5], ' '),

                (
                    str(
                        (
                            (' ' if direction >= 0 else '-')
                            + str(floor(abs(direction)))
                        ).rjust(4, ' ')
                        + '.'
                        + str(
                            int(
                                100 * round(
                                    direction - floor(direction),
                                    2
                                )
                            )
                        ).ljust(2, '0')
                    ) + '\u00b0'
                    if abs(round_state[k]) > 0
                    else offsets[6] * ' '
                ).ljust(offsets[6], ' '),

                fg(*[int(255 * a) for a in complex_to_rgb(state[k])]) + (
                    int(abs(state[k] * 24)) * symbol
                ).ljust(offsets[7], ' ') + fg.rs

                # str(round(abs(state[k]) ** 2, decimals)).ljust(
                #     decimals + 2, ' '
                # ) + bg.rs
            ]]

    sorted_rows = sorted(rows, key=lambda x: int(x[0]))

    for row in sorted_rows:
        output += '  '.join(row)
        output += '\n'

    return output

In [6]:
from math import ceil, log2

from hume import QuantumRegister, QuantumCircuit
from apps.utils import knapsack_table_new
from hume.algos.grover_optimizer import grover_optimizer, oracle_match_0_multi
from hume.algos.function_encoding import encode_term
from hume.utils.common import padded_bin

# Define and run a Grover-based quantum knapsack solver that builds the circuit,
# encodes item values and weights, searches for valid selections, and reports
# the best solution found.

def build_knapsack_circuit(n_key, n_value, n_weight, v, w):

    key = QuantumRegister(n_key)
    value = QuantumRegister(n_value)
    weight = QuantumRegister(n_weight)
    circuit = QuantumCircuit(key, value, weight)

    for i in range(len(key)):
        circuit.h(key[i])

    for i in range(len(value)):
        circuit.h(value[i])

    for i in range(len(weight)):
        circuit.h(weight[i])

    # encode value function
    for (coeff, vars) in v:
        encode_term(coeff, vars, circuit, key, value)

    circuit.iqft(value[::-1], swap=False)

    # encode weight function
    for (coeff, vars) in w:
        encode_term(coeff, vars, circuit, key, weight)

    circuit.iqft(weight[::-1], swap=False)

    return circuit


def terms(l):
    return  [(l[k], [k]) for k in range(len(l))]


def get_selection_value(selection_binary, v):
    val = 0
    for i in range(len(selection_binary)):
        val += int(selection_binary[::-1][i]) * v[i][0]

    return val


def get_selection_weight(selection_binary, w):
    weight = 0
    for i in range(len(selection_binary)):
        weight += int(selection_binary[::-1][i]) * w[i][0]

    return weight


def solve_knapsack(values, weights, max_weight):

    def action_on_progress(flow_state):
        n_key = flow_state['circuit_params']['n_key']
        n_value = flow_state['circuit_params']['n_value']
        n_weight = flow_state['circuit_params']['n_weight']
        s = flow_state['last_run_result']['state vector']
        print(knapsack_table_new(s, n_key, n_value, n_weight))

        print(f'\nLooking for values >= {
            flow_state["circuit_params"]["min_value"]}')


    assert(len(values) == len(weights))
    n_key = len(values)

    v = terms(values)
    w = terms(weights)

    v_adjusted = [(-max(values) - 0, [])] + v
    w_adjusted = [(-item[0], item[1]) for item in w] + [(max_weight, [])]

    min_value = max(values)
    n_value = ceil(log2(max(min_value, sum(values) - min_value))) + 1

    # weight register size based on w_adjusted
    n_weight = ceil(log2(sum([max_weight - item[0] for item in w]))) + 1

    oracle = oracle_match_0_multi(n_key + n_value + n_weight,
                                  [n_key + n_value - 1,
                                   n_key + n_value + n_weight - 1])

    def process_outcome(outcome, flow_state):
        n = flow_state['circuit_params']['n_key'] + \
              flow_state['circuit_params']['n_value'] + \
              flow_state['circuit_params']['n_weight']
        outcome_selection = padded_bin(n, outcome)[-n_key:]
        outcome_value = get_selection_value(outcome_selection, v)
        outcome_weight = get_selection_weight(outcome_selection, w)
        return outcome_selection, outcome_value, outcome_weight

    def update_circuit_params(outcome_results, flow_state):
        circuit_params = flow_state['circuit_params']
        outcome_selection, outcome_value, outcome_weight = outcome_results
        v = circuit_params['v']

        # adjust value function
        # v_adjusted = v + [(-outcome_value - 0, [])]
        v[0] = (-outcome_value - 1, [])

        # circuit_params['v'] = v_adjusted

        circuit_params['min_value'] = outcome_value + 1
        action_on_progress(flow_state)

    def progress(results, flow_state):
        outcome_selection, outcome_value, outcome_weight = results
        min_value = flow_state['circuit_params']['min_value']
        return (outcome_value >= min_value) and (outcome_weight <= max_weight)

    def build_circuit(flow_state):
        return build_knapsack_circuit(flow_state['circuit_params']['n_key'],
                                      flow_state['circuit_params']['n_value'],
                                      flow_state['circuit_params']['n_weight'],
                                      flow_state['circuit_params']['v'],
                                      flow_state['circuit_params']['w'])

    results = grover_optimizer({'n_key': n_key, 'n_value': n_value,
                                'n_weight': n_weight, 'v': v_adjusted,
                                'w': w_adjusted, 'min_value': min_value},
                     build_circuit, oracle,
                     update_circuit_params, progress, process_outcome,
                               failure_threshold=3)

    items_string = results[0]
    items = [str(len(items_string) - 1 - k)
      for k in range(len(items_string)) if items_string[k] == '1'][::-1]

    print(
      f'Optimal selection consists of items {", ".join(items)}. '
      f'The combined value is {results[1]} and the combined weight '
      f'is {results[2]}.'
    )

def test_knapsack():
    values = [5, 3, 1]
    weights = [3, 2, 1]
    max_weight = 5


    solve_knapsack(values, weights, max_weight)

test_knapsack()


iteration 0
failure ('100', 1, 1)

iteration 1
progress ('101', 6, 4)

Selection  Value  Binary  Weight  Binary  Magnitude  Direction  Amplitude Bar           
----------------------------------------------------------------------------------------
000        11     1011    5       00101   0.1768        0.00°   ████                    
001        0      0000    2       00010   0.5303      180.00°   ████████████            
010        14     1110    3       00011   0.1768        0.00°   ████                    
011        3      0011    0       00000   0.5303      180.00°   ████████████            
100        12     1100    4       00100   0.1768        0.00°   ████                    
101        1      0001    1       00001   0.5303      180.00°   ████████████            
110        15     1111    2       00010   0.1768        0.00°   ████                    
111        4      0100    31      11111   0.1768        0.00°   ████                    


Looking for values >= 7

iteration 0

## Try It: Change the Knapsack Parameters

Find the lines:

  values = [2, 3, 1]  
  max_weight = 4  

Change the maximum weight allowed in the knapsack and observe how the optimal solution changes.

Try:

**max_weight = 3**

and then

**max_weight = 5**

Next, modify the item values and observe how the selected items change. For example:

**values = [5, 3, 1]**

What items does the algorithm select in each case, and how do the value and weight constraints affect the result?

<details>
  <summary><strong>Click to reveal expected observation: Changing the maximum weight</strong></summary>
  <p>
    When the maximum allowed weight changes, the set of valid item combinations
    also changes. If the maximum weight is reduced (for example, to 3), some
    combinations that were previously valid become too heavy and are excluded
    from the solution space. If the maximum weight is increased (for example,
    to 5), additional combinations become possible, and the algorithm may find
    a different selection with a higher total value.
  </p>
</details>

<details>
  <summary><strong>Click to reveal expected observation: Changing item values</strong></summary>
  <p>
    Changing the item values alters which combinations produce the highest
    total value. The Grover-based search amplifies states corresponding to
    selections that satisfy the weight constraint and exceed the current
    value threshold. As a result, modifying the values list changes which
    item combinations are considered optimal and therefore which selection
    the algorithm reports as the best solution.
  </p>
</details>

## More on Single-Qubit States and Gates

A quantum state can be represented in code as a state vector consisting of a pair of amplitudes which describe the contribution of each basis state.
- As we saw in the previous workshop session, quantum states are often written as kets using Dirac notation:

  |ψ⟩ = α|0⟩ + β|1)

- e.g.,

  ∣ψ⟩=(-0.34+0.43i)|0⟩+(0.68+0.49i)|1⟩

- The qubit state shown above is described as a superposition of |0> and |1>, where the complex amplitudes determine the measurement probabilities after computing their squared magnitudes.
- For a complex amplitude a+bi, the probability is a^2 + b^2.

  |-0.34+0.43i|^2 = (-0.34)^2 + (0.43)^2 = 0.1156+0.1849 ≈ 0.3005

  |0.68+0.49i|^2 = (0.68)^2 + (0.49)^2 = 0.4624 + 0.2401 ≈ 0.7025

- Final probabilities
  - P(0) ≈ 0.3005
  - P(1) ≈ 0.7025

  - Note that these sum to approximately 1 (because the original amplitudes were rounded to two places, otherwise they would sum to exactly 1), which is required for a valid quantum state.

## Trigonometric Form of a Qubit State
A quantum state of a qubit can be expressed using trigonometric formulas by describing its amplitudes with angles.
- Instead of writing arbitrary complex numbers, the amplitudes are written using sine and cosine functions so that the probabilities automatically sum to 1.
- A common form (known as **Bloch sphere parameterization**) is

  |ψ> = cos(θ/2)|0> + e^(iφ) sin(θ/2)|1>

  or

  “psi equals cosine theta over two times ket zero, plus e to the i phi times sine of theta over two times ket one.”

  - (the constant e is used because it provides a convenient way to represent rotations and oscillations, fundamental to quantum mechanics).
  
  where the angles determine the magnitude and phase of the amplitudes.
- As we saw in the Day 1 workshop, these angles can be represented as a point on the Bloch sphere, allowing quantum states to be visualized as rotations on the sphere.
- We will mostly avoid this notation in this workshop; not because it isn’t important, but because saying things like “e to the i phi times sine of theta over two times ket one” before coffee can cause unnecessary stress.

## Hands-On: A Bloch Sphere Visualization Tool

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/bloch-sphere-visualizer.png"
     alt="Bloch Sphere Visualizer"
     width="25%">

The Bloch Sphere Visualizer at https://bloch.kherb.io/ is an interactive web tool for visualizing the state of a single qubit on a Bloch sphere.
- It provides an intuitive way to see how quantum states change when quantum gates or rotations are applied.
### Description
- Z axis (top and bottom): Represents the computational basis states |0⟩ and |1⟩, which determine the measurement outcomes.
- X axis (left to right across the equator): Represents equal superpositions of |0⟩ and |1⟩ with no phase difference, such as (|0⟩±|1⟩)/√2.
- Y axis (front to back across the equator): Represents equal superpositions where the amplitudes differ by an imaginary phase (±i), such as (|0⟩±i|1⟩)/√2.
### Demonstration
- Rotation Around the X Axis
  - When you rotate around the X axis, the state moves in a vertical circle that passes through the Y direction.
  - Think of it like spinning the sphere around the horizontal rod that goes left to right.
- Start State (INIT)
  - Start at |0⟩, which is the north pole.
- Apply an X-axis rotation of +90° (click +90° on the X axis)
  - The state moves from the north pole toward the front of the sphere, landing on the +Y axis.
    - Mathematically, this is approximately (|0⟩ + i|1⟩)/√2, a superposition with phase where the probabilities of measuring 0 or 1 are equal: P(0) = 0.5 and P(1) = 0.5.
    - This introduces an **i phase**, meaning the amplitude of |1⟩ is multiplied by i (i.e., √-1), creating a 90° phase difference between the amplitudes of |0⟩ and |1⟩.
    - This phase difference places the state on the +Y axis of the Bloch sphere.
    - This does not change measurement probabilities, but it affects how the state interferes with other states during computation.
- Rotate another +90° on the X axis (for a total 180°). Now the state reaches the south pole, or |1⟩
  - The 180° rotation around X flips the qubit (swaps the amplitudes of |0⟩ and |1⟩.
  - This is equivalent to the behavior of a quantum NOT gate (e.g., the **Pauli-X gate**), which flips the state of a qubit.
- Continue rotating another +90°
  - The state becomes (|0⟩ - i|1⟩)/√2, a superposition with equal probabilities but a -90° phase shift.
  - Now we are on the opposite side of the Y axis.
- Rotate a final +90° (360° total), returning to |0⟩.
- In general:
  - Rotation around the X axis mixes amplitudes and introduces phase (flips the qubit through superposition).
  - Rotation around the Y axis mixes amplitudes without introducing complex phase (moves the qubit smoothly between |0⟩ and |1⟩).
  - Rotation around the Z axis changes phase only (spins the phase around the equator while probabilities stay the same).
- Pauli Gates are basic single-qubit quantum gates that perform 180° rotations of a qubit’s state around the principal axes of the Bloch sphere.
  - Pauli-X (X): 180° rotation around the X axis (swaps |0⟩ and |1⟩, similar to a classical NOT gate)
  - Pauli-Y (Y): 180° rotation around the Y axis (swaps |0⟩ and |1⟩ while introducing phase)
  - Pauli-Z (Z): 180° rotation around the Z axis (changes the relative phase between |0⟩ and |1⟩ without changing measurement probabilities)

## Single Qubit with Example Amplitudes
- For a single qubit, there are two possible outcomes (0 and 1), each associated with a complex amplitude.
- These amplitudes determine the probabilities of measurement and can be represented in algebraic form within the table.
- Complex numbers can be visualized on the complex plane, where the horizontal axis represents the real part and the vertical axis represents the imaginary part.
- On this plane, the amplitude’s magnitude corresponds to its distance from the origin, and its direction corresponds to its phase (the angle from the real axis).
- The magnitude determines measurement probability (by squaring it, as we saw above), while the phase  influences how states interfere (combine and interact) during computation.
- The probabilities must sum to 1.
  - => if the probability of outcome 0 is p, then the probability of outcome 1 is 1 - p.
- These values can all be visualized in state tables using bars whose length represents magnitude and color represents direction.

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/figure3.5.png"
      alt="Tabular Representation of Single Qubit State"
      width="75%">

**Tabular Representation of Single Qubit State**

### Polar Form
- A complex number can be represented by a magnitude r and an angle θ (which represents the direction).
- Using this information, we can write the number z = x + iy as
  - x = r cosθ
  - y = r sinθ
- which gives us
  - z = r(cos θ + i * sin θ)
- This is useful in quantum computing because quantum gates often rotate amplitudes by changing their phase θ while keeping their magnitude the same.
  - Many quantum operations are rotations, and rotations are much easier to describe with magnitude + angle than with real/imaginary parts.



## Programmatically Encoding Single-Qubit States with Python Lists
- A single-qubit quantum state can be expressed as a Python list of two amplitudes, where the index of an amplitude is the corresponding outcome.
- Python represents the imaginary unit, i, in a complex number with the symbol j.
  - j is used instead of i because in engineering i usually represents electric current; Python's creator (Guido van Rossum) used this convention as a design choice.
- To encode the state described above:
```
  state = [0.2958+0.51235j, -0.40311+0.69821j]
```
- The index of each amplitude is the corresponding outcome (e.g., state[0], state[1])

In [7]:
# quantum state example

# set the amplitudes
state = [0.2958+0.51235j, -0.40311+0.69821j]
print(state[0])
print(state[1])

# square the amplitudes and print their sum, what are we calculating here?
sqr0 = abs(state[0])**2
sqr1 = abs(state[1])**2
print(sqr0)
print(sqr1)
print(sqr0 + sqr1)

(0.2958+0.51235j)
(-0.40311+0.69821j)
0.3500001625
0.6499948762000001
0.9999950387000001


### On the Bloch Sphere
- This state is not on the north or south pole, so it is in superposition.
- Since we know that on the Bloch sphere,
  - North pole	P(0) = 1
  - Equator	P(0) = P(1) = 0.5
  - South pole	P(1) = 1
- So P(1) > P(0), and we can see that the point is below the equator (more toward 1 than 0)
- Skipping the heavy math, we magically calculate
  - 𝜃 ≈ 107.46 and 𝜙 ≈ -60.00
- To visualize this, use the tool and rotate from |0⟩:
  - Rotate Y by +107°
  - Rotate Z by +60°

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/blochsphere-day2pythonexample1.png"
      alt="Python List Example Bloch Sphere"
      width="75%">

- NOTE: although we rotate +60° around the Z axis in the tool, the Bloch-sphere phase for this state is -60; the simulator uses the opposite rotation convention, so a +60° tool rotation corresponds to a phase of -60°.

## Implementing a Single-Qubit Quantum Simulator

- Start the Qubit in the default state |0>
  - The probability of measuring 0 is 100%, as created by the following init_state following function

  NOTE: a Python **list comprehension** is a compact way to create a new list by applying an expression to each item.

In [8]:
# function to init qubit state to |0>
def init_state():
  # list comprehension to create list of state data
  state = [0 for _ in range(2)]
  # initialize amplitude (1 for outcome 0)
  state[0] = 1
  return state

state = init_state()
# output is a state where amplitude for outcome 0 is 1
# i.e., [amplitude, outcome]
print(state)

[1, 0]


## Hands-On: Preparing a Single Qubit State
- On a real quantum computer, amplitudes cannot be set directly; gates must be used to change them.
- In a classical simulator, however, the states can be set as a list of amplitudes.
- A valid single-qubit state must have two amplitudes whose squared magnitudes sum to 1, which the prepare_state function below checks for.

NOTE: a Python **assert** statement checks that a condition is true during program execution and raises an error if the condition is false


In [9]:
from ch03.util import is_close

def prepare_state(*a):
  # create a new state from the list parameter a
  state = [a[k] for k in range(len(a))]
  # check that the state has 2 amplitudes
  assert(len(state) == 2)
  # check that the sum of squares is close enough to 1
  assert(is_close(sum([abs(state[k])**2 for k in range(len(state))]), 1))
  return state

list = [0.2958+0.51235j, -0.40311+0.69821j]
state = prepare_state(*list)
print(state)

[(0.2958+0.51235j), (-0.40311+0.69821j)]


### Try It: Test an Invalid State
Test the assertion code above by removing the complex identifier (j) from one of the numbers in the list, then rerun the code block.

## Changing amplitudes with single-qubit gates
- After initializing a quantum state, the next step is changing amplitudes using quantum gates.
- Gates are basic instructions that modify pairs of amplitudes in a quantum system.
- For a single qubit, there is only one pair of amplitudes, so gates change their values while preserving total probability.

## Common Single-Qubit Gates
As we have seen, a complex number with magnitude r and angle θ can be written in polar form as r(cosθ+isinθ).
- Multiplying a complex number by another complex number with magnitude 1 rotates the original number without changing its magnitude (changes only the angle, not the length).
- The result is a counterclockwise rotation of the complex number by an angle φ
- We will see that many single-qubit gates work in this way: by rotating the amplitudes of the quantum state in the complex plane.
- Rotations can be implemented in code using cis notation defined by

  **cis φ**

  which is simply a shorthand notation for complex numbers in polar form, i.e., cos(φ) + i sin(φ)

In [12]:
# shortcut function for rotations
from math import cos, sin, pi
def cis(theta):
  return cos(theta) + 1j*sin(theta)

## Basic Single-Qubit Gates
- Some gates change the outcome probabilities, some change the directions, and some change both.

### The X Gate
- Also called the NOT gate, the X gate swaps the amplitudes corresponding to a pair of outcomes.
- The state table below shows the swapped outcomes.
data in the rows corresponding to those outcomes

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/xgate-statetable.png"
      alt="State Table for an X Gate"
      width="75%">

- To simulate an X gate applied to a single-qubit state in code,  just swap the items in the amplitudes:

```Python
  state = [state[1], state[0]]
```

### The Z Gate
- The Z gate multiplies the "1" side of a pair of amplitudes by -1.
- It changes the signs of both the real and the imaginary parts of the amplitude.

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/zgate-statetable.png"
      alt="State Table for an Z Gate"
      width="75%">

- The magnitude of that amplitude does not change (so the  probability does not change), but the direction is reversed by 180°.

```Python
  state = [state[0], -state[1]]
```


In [13]:
# Textbook Exercise 3.2
# Write code that verifies that multiplying the 1 side of the
# following example state by –1 reverses its direction:
# state = [sqrt(0.3), -sqrt(0.7)]

from math import sqrt

# Original state
state = [sqrt(0.3), -sqrt(0.7)]

# Flip the sign of the |1⟩ amplitude
flipped_state = [state[0], -state[1]]

print("Original state:      ", state)
print("After sign flip:     ", flipped_state)

# Compute dot product to compare direction
dot_product = state[0]*flipped_state[0] + state[1]*flipped_state[1]

print("\nDot product:", dot_product)

Original state:       [0.5477225575051661, -0.8366600265340756]
After sign flip:      [0.5477225575051661, 0.8366600265340756]

Dot product: -0.40000000000000013


- The dot product used above is a way to measure how much two vectors point in the same direction.
  - [ a, b ] · [ c, d ] = (a x c) + (b x d)
- Interpretation:
  - Large positive → pointing in the same direction
  - Zero → unrelated / perpendicular
  - Negative → pointing in opposite directions
- In the above exercise, we did not change probabilities; 0.3 and 0.7 stay the same
- We changed the phase/sign, which means the state now points in a different direction in state space.
- -0.4 means the vectors point in different directions, but not completely opposite, demonstrating the state changed direction without fully flipping (which would be a dot product of -1).

### The Phase Gate
- A phase gate P(ϕ) rotates the 1 side of a pair of amplitudes by a given angle ϕ (phi) in radians.
- It changes the “angle” (phase) of a quantum state without changing its probabilities (leaves |0⟩ alone).

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/phasegate-statetable.png"
      alt="State Table for an Phase Gate"
      width="50%">

```Python
  phi = pi/3
  state = [state[0], cis(phi)*state[1]]
```

- On the Bloch sphere, this is a rotation around the Z-axis (the state moves in the complex plane)

In [14]:
state = [sqrt(0.3), -sqrt(0.7)]
phi = pi/3
print(state)
state = [state[0], cis(phi)*state[1]]
print(state)

[0.5477225575051661, -0.8366600265340756]
[0.5477225575051661, (-0.4183300132670379-0.7245688373094719j)]


## The Hadamard Gate
Recall the Hadamard gate from Day 1, which places a qubit into an equal superposition of the states |0> and |1>.
- The transformation maintains |0⟩ and |1⟩ in two distinct states (keeps them orthogonal so we can still tell them apart).
- Hadamard is reversible because it preserves enough information (via phase) to recover the original state.
- Applied to the |0> state
  - H|0> = (|0> + |1>) / √2
- applied to the |1> state
  - H|1> = (|0> - |1>) √2

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/hgate-statetable.png"
      alt="State Table for a Hadamard Gate"
      width="50%">

- This gate is commonly used at the beginning of many quantum algorithms to o create a superposition representing all possible candidate solutions simultaneously.
- To simulate the gate:
```Python
  state = [sqrt(0.5)*(state[0] + state[1]), sqrt(0.5)*(state[0] - state[1])]
```

## The Rz Gate
- Rz(θ) rotates the phases of the amplitudes: the |0⟩ part rotates one way and the |1⟩ part rotates the opposite way (by θ/2 each).
- It changes phase but not probabilities, so measurement outcomes stay the same while interference behavior changes.

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/rzgate-statetable.png"
      alt="State Table for an Rz Gate"
      width="50%">

```Python
theta = pi/3
state = [cis(-theta/2)*state[0], cis(theta/2)*state[1]]
```

## The Y Gate
- The Y gate rotates phases: the |0⟩ part by +90° and the |1⟩ part by -90°.
- Then it swaps the amplitudes of |0⟩ and |1⟩.
- This results in the probabilities being swapped and a change in phase (affecting interference)

<img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/ygate-statetable.png"
      alt="State Table for a Y Gate"
      width="50%">

```Python
state = [-1j*state[1], 1j*state[0]]
```

## More Basic Single-Qubit Gates
**Rx** (Rotation around X-axis)
- Rotates the qubit around the X-axis of the Bloch sphere by angle θ
- Mixes amplitudes (changes both magnitude and phase)
- Introduces complex (imaginary) components
- Useful for creating states with phase differences
- Can move a state out of the Z-axis into the Y direction

**Ry** (Rotation around Y-axis)
- Rotates the qubit around the Y-axis of the Bloch sphere by angle θ
- Adjusts amplitudes without introducing complex numbers
- Keeps amplitudes real-valued
- Acts like a "biased coin dial" between |0> and |1>

### Quick Contrast
- Rx: changes phase and probability (complex amplitudes).
  - Use it when you need phase manipulation for interference effects (key in quantum algorithms).
- Ry: changes probability only (real amplitudes).
  - Use it when you want intuitive control over measurement outcomes, e.g., applying Ry(π/2) to |0> creates an equal superposition.

## Where Do Rotation Names Come From?
- Rotations are standard in controlling the flight of an object, as they allow it to maneuver in the air and change its direction.
- The image from our textbook shown below provides a visual analogy for understanding rotations, but in quantum computing these rotations act on a qubit’s state (probability and phase on the Bloch sphere) rather than a physical object in space.

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/rotation-names.png"
        alt="Rotation Names"
        width="50%">

- In the image, rotations are shown along of three axes: **roll (X)**, **pitch (Y)**, and **yaw (Z)**.
  - Roll is the rotation around the X, or longitudinal (front-to-back) axis and is the movement of the object banking left or right.
  - Pitch is the rotation around the Y, or lateral (side-to-side) axis and is the movement of the object’s front up or down.
  - Yaw is the rotation around the Z, or vertical (up-and-down) axis, and is the movement of the object’s nose from side to side.
- Instead of using the terms "roll", "pitch", and "yaw", we typically just say "rotation around X, or Rx(θ); rotation around Y, or Ry(θ), and rotation around Z, or Rz(θ).
- Remember these three key points about rotation:
  - Rx rotation moves state around the sphere horizontally
  - Ry rotation moves state up/down between |0⟩ and |1⟩
  - Rz rotation spins the state around itself (phase change)

## Single-Qubit Gate Inverses
- All quantum gates are reversible, meaning they have an inverse operation that can undo their effect.
- Some gates are self-inverse, meaning they undo themselves:
  - The X, Y, Z, and H gates all satisfy X² = I, Y² = I, Z² = I, and H² = I, meaning each gate is its own inverse.
  - Squaring (reapplying) these operations results in the **identity matrix** which means they "cancel themselves out" (the second application reverses the first).
```
      I =
        [1, 0]  
        [0, 1]
```
  - In the textbook you will see this represented using **U**, a general symbol for a quantum gate (unitary operator), allowing properties like inverses to be expressed in a standard, gate-independent way (e.g., U² = I).
- Other gates are reversed by applying the opposite operation:
  - For example, rotations:
    - Rz(θ)<sup>-1</sup> = Rz(-θ)
- Inverse notation is written using a power of -1:
  - e.g., X<sup>-1</sup> = X


## Example: Demonstrate that the X gate is its own inverse

- In the example shown below, the result of X @ X is:

```
    [1, 0]
    [0, 1]
```

- **np.allclose(X @ X, I)** checks whether two arrays are approximately equal, avoiding issues that arise when using == with floating-point values due to small rounding errors.
- It returns:
  - True → if all corresponding values are “close enough”
  - False → if they differ beyond a small tolerance
- The check returns True, which confirms that X² = I, meaning the X gate is self-inverse
- Note: The textbook represents qubit states as simple lists of amplitudes (e.g., [a, b]), while the example above uses a matrix representation (NumPy array) for gates and operations; both describe the same concepts; matrices can make it easier to demonstrate operations like gate multiplication. Just remember:
  - A qubit is a vector
  - A gate is a matrix
  - Applying a gate = matrix x vector

In [15]:
import numpy as np

# Define the X gate
X = np.array([
    [0, 1],
    [1, 0]
], dtype=complex)

# Identity matrix
I = np.eye(2, dtype=complex)

# Show the gate
print("X =")
print(X)

# Verify X is its own inverse
print("\nX @ X =")
print(np.round(X @ X, 5))
print("Is X its own inverse?", np.allclose(X @ X, I))

X =
[[0.+0.j 1.+0.j]
 [1.+0.j 0.+0.j]]

X @ X =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Is X its own inverse? True


## Hands-On: Verifying Self-Inverse Gates
- Work through the following fill-in-the-blank exercise. The solution follows, but try to complete it yourself first.
- Complete the matrix definitions for the Y, Z, and H gates
- Run the code to verify that squaring each gate results in the identity matrix

In [21]:
import numpy as np

# Define the Y, Z, and H gates
Y = np.array([
    [0, 1],
    [1, 0]
], dtype=complex)
# Hint: Y uses imaginary values (±i);
# the top-right is −1i and the bottom-left is +1i
# (but remember to use Python’s notation for i)

Z = np.array([
    [1, 0],
    [0, 1]
], dtype=complex)
# Hint: Z applies a phase of −1 to the |1⟩ state;
# the diagonal entries are 1 (for |0⟩) and −1 (for |1⟩)

H = (1 / np.sqrt(2)) * np.array([
    [1, 1],
    [1, -1]
], dtype=complex)
# Hint: H creates equal superposition with opposite signs;

# Identity matrix
I = np.eye(2, dtype=complex)

# Show the gates clearly
print("Y =")
print(Y)
print("\nY @ Y =")
print(np.round(Y @ Y, 5))
print("Is Y its own inverse?", np.allclose(Y @ Y, I))
print()

print("Z =")
print(Z)
print("\nZ @ Z =")
print(np.round(Z @ Z, 5))
print("Is Z its own inverse?", np.allclose(Z @ Z, I))
print()

print("H =")
print(np.round(H, 5))
print("\nH @ H =")
print(np.round(H @ H, 5))
print("Is H its own inverse?", np.allclose(H @ H, I))

Y =
[[0.+0.j 1.+0.j]
 [1.+0.j 0.+0.j]]

Y @ Y =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Is Y its own inverse? True

Z =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]

Z @ Z =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Is Z its own inverse? True

H =
[[ 0.70711+0.j  0.70711+0.j]
 [ 0.70711+0.j -0.70711+0.j]]

H @ H =
[[ 1.+0.j -0.+0.j]
 [-0.+0.j  1.+0.j]]
Is H its own inverse? True


In [22]:
import numpy as np

# Define the Y, Z, and H gates
Y = np.array([
    [0, -1j],
    [1j, 0]
], dtype=complex)

Z = np.array([
    [1, 0],
    [0, -1]
], dtype=complex)

H = (1 / np.sqrt(2)) * np.array([
    [1, 1],
    [1, -1]
], dtype=complex)

# Identity matrix
I = np.eye(2, dtype=complex)

# Show the gates clearly
print("Y =")
print(Y)
print("\nY @ Y =")
print(np.round(Y @ Y, 5))
print("Is Y its own inverse?", np.allclose(Y @ Y, I))
print()

print("Z =")
print(Z)
print("\nZ @ Z =")
print(np.round(Z @ Z, 5))
print("Is Z its own inverse?", np.allclose(Z @ Z, I))
print()

print("H =")
print(np.round(H, 5))
print("\nH @ H =")
print(np.round(H @ H, 5))
print("Is H its own inverse?", np.allclose(H @ H, I))

Y =
[[ 0.+0.j -0.-1.j]
 [ 0.+1.j  0.+0.j]]

Y @ Y =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Is Y its own inverse? True

Z =
[[ 1.+0.j  0.+0.j]
 [ 0.+0.j -1.+0.j]]

Z @ Z =
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]
Is Z its own inverse? True

H =
[[ 0.70711+0.j  0.70711+0.j]
 [ 0.70711+0.j -0.70711+0.j]]

H @ H =
[[ 1.+0.j -0.+0.j]
 [-0.+0.j  1.+0.j]]
Is H its own inverse? True


## Transforming a Single Qubit State
- Now let's do some more experimentation with gate operations.
- First let's define functions to represent the gates
  - (but let's be good developers by including descriptive comments).



In [25]:
from math import cos, sin, atan2, sqrt, pi
import numpy as np

# X gate
# Flips the qubit state: |0⟩ ↔ |1⟩
x = [
    [0, 1],
    [1, 0]
]

# Z gate
# Leaves |0⟩ unchanged and flips the phase of |1⟩
z = [
    [1, 0],
    [0, -1]
]

# Phase gate
# Applies a phase rotation to the |1⟩ amplitude by angle theta
def phase(theta):
    return [
        [1, 0],
        [0, complex(cos(theta), sin(theta))]
    ]

# Hadamard gate (H)
# Creates superposition: maps |0⟩ → (|0⟩ + |1⟩)/√2
h = [
    [1/sqrt(2),  1/sqrt(2)],
    [1/sqrt(2), -1/sqrt(2)]
]

# Rz rotation gate
# Rotates the qubit around the Z axis of the Bloch sphere by angle theta
def rz(theta):
    return [
        [complex(cos(theta / 2), -sin(theta / 2)), 0],
        [0, complex(cos(theta / 2),  sin(theta / 2))]
    ]

# Y gate
# Combines a bit flip with a phase change (uses imaginary values ±i)
y = [
    [0, complex(0, -1)],
    [complex(0, 1), 0]
]

# Rx rotation gate
# Rotates the qubit around the X axis by angle theta
def rx(theta):
    return [
        [cos(theta/2), complex(0, -sin(theta/2))],
        [complex(0, -sin(theta/2)), cos(theta/2)]
    ]

# Ry rotation gate
# Rotates the qubit around the Y axis by angle theta
def ry(theta):
    return [
        [cos(theta/2), -sin(theta/2)],
        [sin(theta/2),  cos(theta/2)]
    ]

# Textbook's transform function
# Apply a 2x2 quantum gate matrix to a single-qubit state vector,
# update amplitudes via matrix multiplication
#
# see below for a cleaner version using @
def transform_book(state, gate):
    assert(len(state) == 2)
    z0 = state[0]
    z1 = state[1]
    state[0] = gate[0][0]*z0 + gate[0][1]*z1
    state[1] = gate[1][0]*z0 + gate[1][1]*z1

# Cleaner version vs. book using numpy's @ operator
def transform(state, gate):
    assert len(state) == 2
    result = np.array(gate) @ np.array(state)
    state[0], state[1] = result


# Convert quantum state vector into table of amplitudes,
# phases in degrees, magnitudes, and probabilities
def to_table(s, decimals=5):
    table = [
        [
            k,
            s[k],
            atan2(s[k].imag, s[k].real) / (2 * pi) * 360,
            abs(s[k]),
            abs(s[k]) ** 2,
        ]
        for k in range(len(s))
    ]

    table_r = [
        [
            float(round(x, decimals))
            if isinstance(x, (float, np.floating))
            else complex(
                round(x.real, decimals),
                round(x.imag, decimals),
            )
            if isinstance(x, (complex, np.complexfloating))
            else x
            for x in row
        ]
        for row in table
    ]

    return table_r

# function to init qubit state to |0>
def init_state():
  # list comprehension to create list of state data
  state = [0 for _ in range(2)]
  # initialize amplitude (1 for outcome 0)
  state[0] = 1
  return state

def print_state(state, decimals=5):
  print(*to_table(state, decimals),sep='\n')

# Create a circuit by applying a sequence of gates
# to transform amplitudes and phase of qubit s

# initialize the state
s = init_state()
# apply gates in order
print_state(s)
# Ry(2π/3) → X → P(π/3) → H    read right to left because it is gates and is in reverse order of operation
transform(s, ry(2*pi/3))
transform(s, x)
transform(s, phase(pi/3))
transform(s, h)
print_state(s)

#binary, amplitude, direction, magnitude, probability


[0, 1, 0.0, 1, 1]
[1, 0, 0.0, 0, 0]
[0, (0.78915+0.30619j), 21.20602, 0.84647, 0.71651]
[1, (0.4356-0.30619j), -35.10391, 0.53244, 0.28349]


## Single Qubit Circuits
- The above code creates a single-qubit circuit by applying a sequence of gates to transform the qubit’s amplitudes.
- These transformations reshape both probability and phase which ultimately determine measurement outcomes.
### Circuit Expressions vs. Diagrams
- **Circuit expressions (gates) are written right-to-left** (reverse of execution order):
  - HP(π/3) X Ry(2π/3)
  - Juxtaposing the gates (writing them next to each other) implies multiplication/composition; operator would make the expressions harder to read.
- Diagrams go left to right in execution order, which is more intuitive visually, but the math is written the opposite way.
- Diagrams are visualized as a horizontal wire (qubit) with gates shown as boxes from left to right in execution order.

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/figure3.33.png"
        alt="Circuit Diagram for Single Qubit"
        width="50%">

### State Table Representation of Results
- After applying the gates, the qubit ends in a new state with complex amplitudes for outcomes 0 and 1.
- A state table summarizes the result, showing for each outcome:
complex amplitude (magnitude + phase)
  - direction (angle)
  - magnitude
  - probability
  - Measurement probabilities are derived from amplitudes:
    - Outcome 0 ≈ 72%
    - Outcome 1 ≈ 28%

  <img src="https://raw.githubusercontent.com/FSCJ-FacultyDev/WC26-SPII-QuantumWorkshop/main/images/figure3.34.png"
          alt="State Table for Single Qubit"
          width="75%">




## Real Hardware vs. Simulation

- Recall that quantum circuits are executed multiple times (shots) on real hardware
  - Each run ends with a measurement step
  - Measurement causes the quantum state to collapse to a single outcome
  - The observed result corresponds to the amplitude that becomes 1 after collapse
  - Individual runs may differ, but counts across many runs show the probability distribution

- Simulation, on the other hand, uses probabilities directly to mimic measurement outcomes
  - No need to rerun the full circuit each time; sampling replaces repeated execution
  - Outcomes are generated by randomly selecting one of the  measurement results based on probability weights (e.g., 0 or 1 for a single qubit)
  - This produces results that approximate the same distribution as real quantum hardware
  - This is faster and more efficient for small circuits, but it does not capture true quantum effects like noise or hardware constraint
  - The selection is weighted by their probabilities (e.g., 72% chance of 0, 28% chance of 1)
  - Each sample picks one outcome according to those weights, mimicking a real measurement

## Probabilistic Sampling
- We can use Python’s **random.choices** function to simulate quantum We can use Python’s random.choices function to simulate quantum measurement by sampling outcomes based on their probabilities
- Probabilities are computed as squared magnitudes of amplitudes:
```
    abs(s[k])**2
```
- This generates a list of measurement results (e.g., 10 samples → [0, 0, 0, 1, ...])
- **Counter** is used to count how often each outcome occurs, e.g., 0 -> 7, 1 -> 3
  - Counter is a Python class that counts elements in an iterable and returns their frequencies
- Small sample sizes approximate the distribution but can vary considerably
- Larger sample sizes (e.g., 1000) produce frequencies closer to true probabilities (e.g., 0 -> 726, 1 -> 274)
- This demonstrates how repeated measurements reveal the underlying probability distribution of a quantum state

In [ ]:
# simulate quantum measurement using probabilistic sampling

from random import choices
from collections import Counter

# 's' is a state vector of arbitrary amplitudes
s = [0.85+0j, 0.53+0j]

# Compute probabilities from amplitudes
probs = [abs(s[k])**2 for k in range(len(s))]

# --- Simulate 10 measurements ---
samples = choices(range(len(s)), probs, k=10)

print("Samples (10 shots):")
print(samples)

# Count outcomes
print("\nCounts (10 shots):")
for k, v in Counter(samples).items():
    print(f"{k} -> {v}")

# --- Simulate 1000 measurements ---
samples = choices(range(len(s)), probs, k=1000)

print("\nCounts (1000 shots):")
for k, v in Counter(samples).items():
    print(f"{k} -> {v}")

## Encoding a Uniform Distribution in a Qubit

- A two-outcome uniform distribution models a fair coin, where each outcome is equally likely..

### Hands-On: Create a “Fair Coin” with a Qubit

- Build a single-qubit state where outcomes 0 and 1 are equally likely, then simulate measurements.
- Enter your code in the code block below and execute it (copy and paste the "fill-in-the-blank" steps below if desired).
- Note: be sure to run the code block in the section titled "Transforming a Single Qubit State", above, before executing so the necessary functions are defined.

**1. Prepare the State**

```
# initialize the qubit in |0⟩
state = __________()

# apply a gate that creates equal probabilities
transform(state, __________)

# display the resulting state
__________(state)
```

- Expected: the amplitudes should be approximately equal (0.70711 and 0.70711). The probabilities should be 0.5 and 0.5.

**2. Compute Probabilities**

```
# compute probabilities from amplitudes
probs = [round(float(abs(state[k])**2), 5) for k in range(len(state))]

print(probs)
```

- Check: do both probabilities look equal?

**3. Simulate Measurements (10 shots)**

```
from random import choices
from collections import Counter

# Simulate 10 measurements
samples = choices(__________, probs, k=__________)

print("Samples:")
print(samples)

# Count outcomes
print("\nCounts:")
for (k, v) in Counter(samples).items():
    print(str(k) + " -> " + str(v))
```

- Check: are the counts exactly equal? Should they be?

**4. Increase the Number of Shots**

- Modify the code to use 1000 shots.

```
samples = choices(range(len(state)), probs, k=__________)

for (k, v) in Counter(samples).items():
    print(str(k) + " -> " + str(v))
```

- Check: do the counts get closer to 50/50?

### Reflection Questions
1. How does the Hadamard gate create equal probabilities?
2. Why aren’t the counts exactly equal for 10 shots?
3. What happens as the number of shots increases?


In [ ]:
from random import choices
from collections import Counter

... enter your code here ...

## Solution

In [ ]:
from random import choices
from collections import Counter

# initialize the qubit in |0⟩
state = init_state()

# apply a gate that creates equal probabilities
transform(state, h)

# display the resulting state
print_state(state)

# compute probabilities from amplitudes
probs = [round(float(abs(state[k])**2), 5) for k in range(len(state))]

print(probs)

# simulate 10 measurements
samples = choices(range(len(state)), probs, k=10)

print("Samples:")
print(samples)

# count outcomes
print("\nCounts:")
for (k, v) in Counter(samples).items():
    print(str(k) + " -> " + str(v))

# simulate 1000 measurements
samples = choices(range(len(state)), probs, k=1000)

print("\nCounts (1000 shots):")
for (k, v) in Counter(samples).items():
    print(str(k) + " -> " + str(v))

### Reflection Questions
- How does the Hadamard gate create equal probabilities?
  - It transforms the qubit into a superposition with equal amplitudes, so squaring them produces equal probabilities (0.5 and 0.5).
- Why aren’t the counts exactly equal for 10 shots?
  - Measurement is random, so with a small sample size, the outcomes can vary due to chance.
- What happens as the number of shots increases?
  - The measured frequencies get closer to the true probabilities, approaching a 50/50 distribution.

## Homework: Create a “Biased Quantum Coin”

- Build a single-qubit state where outcomes 0 and 1 are not equally likely, then simulate measurements.
- Enter your code in the code block below and execute it.
- Note: be sure to run the code block in the section titled "Transforming a Single Qubit State", above, before executing so the necessary functions are defined.

In [ ]:
# 0. import the required modules
# 1. Initialize the qubit in the |0⟩ state
# 2. Apply one or more gates to create unequal probabilities
#    by using a rotation or combination of gates.
#    Hint: Try using a rotation gate like ry(θ); different angles change
#    how much probability shifts toward |0⟩ or |1⟩ (experiment with θ).
# 3. Display the resulting state to observe amplitudes
# 4. Compute probabilities from amplitudes (squared magnitudes)
# 5. Simulate 10 measurements using probabilistic sampling
# 6. Count how many times each outcome appears (10 shots)
# 7. Simulate 1000 measurements
# 8. Count how many times each outcome appears (1000 shots)
# 9. Verify that one outcome occurs more frequently than the other

## References

- Cho, A. (2023, February 22). Quantum computers take key step toward curbing errors. Science. https://www.science.org/content/article/quantum-computers-take-key-step-toward-curbing-errors
- Gonciulea, C., & Stefanski, C. (2025). Building Quantum Software With Python: A Developer’s Guide. Manning Publications. ISBN: 978-1633437630.